core/src/banking_stage/forward_packet_batches_by_accounts.rs

```diff
@@ -104,39 +74,86 @@ impl ForwardPacketBatchesByAccounts {
 
     pub fn new(limit_ratio: u32, number_of_batches: u32) -> Self {
         let forward_batches = (0..number_of_batches)
-            .map(|_| ForwardBatch::new(limit_ratio))
+            .map(|_| ForwardBatch::default())
             .collect();
-        Self { forward_batches }
+
+        let batch_vote_limit = block_cost_limits::MAX_VOTE_UNITS.saturating_div(limit_ratio as u64);
+        let batch_block_limit =
+            block_cost_limits::MAX_BLOCK_UNITS.saturating_div(limit_ratio as u64);
+        let batch_account_limit =
+            block_cost_limits::MAX_WRITABLE_ACCOUNT_UNITS.saturating_div(limit_ratio as u64);
+
+        let mut cost_tracker = CostTracker::default();
+        cost_tracker.set_limits(
+            batch_account_limit.saturating_mul(number_of_batches as u64),
+            batch_block_limit.saturating_mul(number_of_batches as u64),
+            batch_vote_limit.saturating_mul(number_of_batches as u64),
+        );
+        Self {
+            forward_batches,
+            cost_tracker,
+            batch_vote_limit,
+            batch_block_limit,
+            batch_account_limit,
+        }
     }
 
-    /// packets are filled into first available 'batch' that have space to fit it.
     pub fn try_add_packet(
         &mut self,
         sanitized_transaction: &SanitizedTransaction,
         immutable_packet: Arc<ImmutableDeserializedPacket>,
         feature_set: &FeatureSet,
     ) -> bool {
-        for forward_batch in self.forward_batches.iter_mut() {
-            if forward_batch
-                .try_add(sanitized_transaction, immutable_packet.clone(), feature_set)
-                .is_ok()
-            {
-                return true;
+        let tx_cost = CostModel::calculate_cost(sanitized_transaction, feature_set);
+
+        if let Ok(updated_costs) = self.cost_tracker.try_add(&tx_cost) {
+            let batch_index = self.get_batch_index_by_updated_costs(&tx_cost, &updated_costs);
+
+            if let Some(forward_batch) = self.forward_batches.get_mut(batch_index) {
+                forward_batch.forwardable_packets.push(immutable_packet);
+            } else {
+                // A successfully added tx_cost means it does not exceed block limit, nor vote
```

Comment:
get_batch_index_by_updated_costs doesn't seem to mod?

Reply:
Updated the comment - mistakenly thought to use "reminder" at first, then changed to divide but forgot update the comment. Thanks for pointing it out.